### PR TITLE
stylo: Use atoms as the pseudo-element back-end.

### PR DESCRIPTION
**Pull request by emilio**
A bit of work left, and we can uber-optimize this (left comments, will fill follow-ups), but this is a decent refactor so I thought I'd rather get some feedback on it.

r? @bholley (not formally yet, maybe, but some feedback is appreciated).

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] These changes do not require tests because geckolib only...

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12815)

### CODE REVIEW

File: ports/geckolib/string_cache/lib.rs [hunk 1]
```diff
@@ -135,7 +139,12 @@ impl Atom {
         let atom = Atom(WeakAtom::new(ptr));
         callback(&atom);
         mem::forget(atom);
-     }
+    }
+
+    #[inline]
+    pub unsafe fn from_static(ptr: *mut nsIAtom) -> Self {
```
Comment (bholley) [inline ports/geckolib/string_cache/lib.rs]: Please document this.

File: ports/geckolib/binding_tools/regen.py [hunk 2]
```diff
@@ -212,6 +220,17 @@ def build(objdir, target_name, kind_name=None,
...
+    if "custom_build" in current_target:
+        print("[CUSTOM] {}::{} in \"{}\"... ".format(target_name, kind_name, objdir), end='')
+        sys.stdout.flush()
+        ret = current_target["custom_build"](objdir, verbose=True)
+        if ret != 0:
+            print("FAIL")
+        else:
+            print("OK")
+
+        return 0
```
Comment (emilio) [inline ports/geckolib/binding_tools/regen.py]: whoops, `return ret`.

File: ports/geckolib/binding_tools/regen_atoms.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    TYPE = "nsICSSPseudoElement"
+
+
+class CSSAnonBoxesAtomSource:
+    PATTERN = re.compile('^CSS_ANON_BOX\((.+),\s*"(.*)"\)')
+    FILE = "dist/include/nsCSSAnonBoxList.h"
+    CLASS = "nsGKAtoms"
+    TYPE = "nsICSSAnonBoxPseudo"
+
+    @staticmethod
```
Comment (emilio) [inline ports/geckolib/binding_tools/regen_atoms.py]: This is not needed anymore.

File: ports/geckolib/string_cache/lib.rs [hunk 4]
```diff
@@ -214,7 +223,7 @@ impl fmt::Debug for Atom {
 
 impl fmt::Display for Atom {
     fn fmt(&self, w: &mut fmt::Formatter) -> fmt::Result {
-        for c in char::decode_utf16(self.as_slice().iter().cloned()) {
+        for c in self.chars() {
```
Comment (bholley) [inline ports/geckolib/string_cache/lib.rs]: I actually wanted @SimonSapin to remove this in #12571, because it makes it too easy to accidentally trigger a slow stringification. Can you revert this line and remove |pub fn chars|? I'd also be ok with chars() being non-public with a comment indicating that it's slow and shouldn't be exposed.

File: components/style/gecko_selector_impl.rs [hunk 5]
```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
...
-    MozTreeCheckbox,
-    MozTreeProgressMeter,
-    MozTreeDropFeedback,
-
-    MozSVGMarkerAnonChild,
-    MozSVGOuterSVGAnonChild,
-    MozSVGForeignContent,
-    MozSVGText,
+// TODO: automatize this list from
+// http://searchfox.org/mozilla-central/source/layout/style/nsCSSPseudoElements.h and
```
Comment (bholley) [inline components/style/gecko_selector_impl.rs]: This gets fixed up in future patches, right?

File: components/style/gecko_selector_impl.rs [hunk 6]
```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
...
+        self.1
+    }
+
+    #[inline]
+    fn from_atom_unchecked(atom: Atom, is_anon_box: bool) -> Self {
+        if cfg!(debug_assertions) {
+            match Self::from_atom(&*atom, true) {
+                Some(pseudo) => {
+                    assert_eq!(pseudo.is_anon_box(), is_anon_box);
+                    return pseudo;
```
Comment (bholley) [inline components/style/gecko_selector_impl.rs]: Why do we need this return?

File: components/style/gecko_selector_impl.rs [hunk 7]
```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
...
+        if iter.next().map(|c| *c) != Some(':' as u16) {
+            return None
+        }
+
+        // This type is just needed because a plain iter.map() doesn't derive
+        // Clone because the closure is not Clone.
+        #[derive(Clone)]
+        struct Utf16toASCIIIter<'a>(Iter<'a, u16>);
+
+        impl<'a> Iterator for Utf16toASCIIIter<'a> {
```
Comment (bholley) [inline components/style/gecko_selector_impl.rs]: I don't understand why we need to do all this junk, though it looks like it goes away in subsequent patches? In the mean time, wouldn't we get the same effect by just comparing against atom!(":before") like from_bytes_iter does? Seems preferable to all that nasty unicode munging (which I didn't really look at because it goes away).

Comment (bholley) [inline components/style/gecko_selector_impl.rs]: (Would still prefer to not introduce the nasty stuff at all though, even for a temporary patch)

File: components/style/gecko_selector_impl.rs [hunk 8]
```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
...
+                let bytes = $pseudo_str.as_bytes();
+                if byte_len == bytes.len() &&
+                    bytes.iter().zip(iter.clone()).all(|(a, b)| b.eq_ignore_ascii_case(a)) {
+                    return Some(PseudoElement($atom, $is_anon_box))
+                }
+            }}
+        }
+
+        parse!("before", atom!(":before"));
+        parse!("after", atom!(":after"));
```
Comment (bholley) [inline components/style/gecko_selector_impl.rs]: Do I understand correctly that this temporarily regresses our support for anon boxes during this patch? I don't see us passing |true| to parse! anywhere. I guess I'm ok with that, though generally would prefer to review less intermediate code...

File: components/style/gecko_selector_impl.rs [hunk 9]
```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
...
+        let bytes = s.as_bytes();
+        let len = bytes.len();
+        Self::from_bytes_iter(bytes.iter().cloned(), in_ua, len)
+    }
+
+    fn from_bytes_iter<I: Iterator<Item=u8> + Clone>(iter: I,
+                                                     in_ua_stylesheet: bool,
+                                                     byte_len: usize) -> Option<Self> {
+        use std::ascii::AsciiExt;
+        macro_rules! parse {
```
Comment (bholley) [inline components/style/gecko_selector_impl.rs]: Per IRC discussion this macro should go away.

File: ports/geckolib/binding_tools/regen_atoms.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    return ident
+
+
+class Atom:
+    def __init__(self, source, ident, value):
+        self.ident = "{}_{}".format(source.CLASS, ident)
+        self.value = value
+        self.cpp_class = source.CLASS
+        # FIXME(emilio): This is crap, and we really could/should just store a
+        # reference to the source, but...
```
Comment (bholley) [inline ports/geckolib/binding_tools/regen_atoms.py]: Why not do that? :-)

File: ports/geckolib/binding_tools/regen_atoms.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+/*
+ * This file contains a helper macro invocation to aid Gecko's style system
+ * pseudo-element integration.
+ *
+ * This file is NOT INTENDED to be compiled as a standalone module.
+ *
+ * Expected usage is as follows:
+ *
+ * Also, it guarantees the property that normal pseudo-elements are processed
+ * before anonymous boxes.
```
Comment (bholley) [inline ports/geckolib/binding_tools/regen_atoms.py]: Nit: put this above the 'expected usage' part.

File: ports/geckolib/glue.rs [hunk 12]
```diff
@@ -286,13 +264,8 @@ pub extern "C" fn Servo_GetComputedValuesForAnonymousBox(parent_style_or_null: *
...
     data.flush_stylesheets();
 
-    let pseudo = match pseudo_element_from_atom(pseudo_tag, /* ua_stylesheet = */ true) {
-        Ok(pseudo) => pseudo,
-        Err(pseudo) => {
-            warn!("stylo: Unable to parse anonymous-box pseudo-element: {}", pseudo);
-            return ptr::null_mut();
-        }
-    };
+    let atom = unsafe { Atom::from_static(pseudo_tag) };
```
Comment (bholley) [inline ports/geckolib/glue.rs]: This is a nice optimization, but a bit scary, since it would lead to an unpaired Release() if we ever got it wrong (this only works because ReleaseAtom() is a no-op for static atoms).

Can we make the gecko Atom struct carry an extra bit that indicates whether the atom was created with from_static, and skip the ReleaseAtom() call in that case? That would reduce the risk here, and also save an FFI call.

File: ports/geckolib/string_cache/lib.rs [hunk 13]
```diff
@@ -117,8 +117,11 @@ impl WeakAtom {
 
     #[inline]
     pub fn eq_str_ignore_ascii_case(&self, s: &str) -> bool {
+        let bytes = s.as_bytes();
         unsafe {
-            Gecko_AtomEqualsUTF8IgnoreCase(self.as_ptr(), s.as_ptr() as *const _, s.len() as u32)
+            Gecko_AtomEqualsUTF8IgnoreCase(self.as_ptr(),
+                                           bytes.as_ptr() as *const _,
+                                           bytes.len() as u32)
```
Comment (bholley) [inline ports/geckolib/string_cache/lib.rs]: yikes. @Manishearth, any ideas on tooling to prevent us from making errors like this in the future?

File: ports/geckolib/glue.rs (hunk 12 above)
Comment (emilio) [inline ports/geckolib/glue.rs]: It's probably better if we generate accessors for the atom bitfield, with the extra benefit that gives us the length (which means in some cases we don't need to make an ffi call to make some comparisons).

Comment (emilio) [inline ports/geckolib/glue.rs]: Oh, also, note that `from_atom_unchecked` still checks the atom in debug builds.

File: components/style/properties/gecko.mako.rs [hunk 14]
```diff
@@ -1339,6 +1339,80 @@ fn static_assert() {
     ${impl_coord_copy('column_width', 'mColumnWidth')}
 </%self:impl_trait>
 
+<%self:impl_trait style_struct_name="Counters"
```
Comment (bholley) [inline components/style/properties/gecko.mako.rs]: Nit: The comment message should say "the content property"

Comment (bholley) [inline components/style/properties/gecko.mako.rs]: Actually, could you split this commit out into a separate PR? I've got to run pretty soon and may not have time to review the whole thing, and it's pretty separate from the atom stuff (and also presumably needs some gecko-side changes).

**Comment by bholley** [comment]
r=me with comments addressed on everything but the |content| patch, which I'd like to see in a separate PR.

@bors-servo delegate+

File: ports/geckolib/string_cache/lib.rs (hunk 4 above)
Comment (SimonSapin) [inline ports/geckolib/string_cache/lib.rs]: I’ve removed `PartialEq<str> for Atom` because, I agree, it’s easy to accidentally use it instead of fast atom comparison.

But consuming `Atom::chars` is O(n) like `str::chars`, and calling it has to be explicit, so I don’t see what the problem is. When possible, it’s better than using `with_str` which allocates.

File: components/style/gecko_selector_impl.rs [hunk 15]
```diff
@@ -12,104 +12,85 @@ use stylesheets::Stylesheet;
...
-    MozRangeTrack,
-    MozRangeProgress,
-    MozRangeThumb,
-    MozMeterBar,
-    MozPlaceholder,
-    MozColorSwatch,
-
-    AnonBox(AnonBoxPseudoElement),
-}
+pub struct PseudoElement(Atom, bool);
```
Comment (SimonSapin) [inline components/style/gecko_selector_impl.rs]: Maybe use struct syntax with named fields? I particular to say what this `bool` is.

File: ports/geckolib/string_cache/lib.rs (hunk 13 above)
Comment (SimonSapin) [inline ports/geckolib/string_cache/lib.rs]: I think this change is a no-op: `str::len` returns a number of UTF-8 bytes, not of code points.

File: components/style/gecko_selector_impl.rs (hunk 15 above)
Comment (emilio) [inline components/style/gecko_selector_impl.rs]: It's documented above (though the diff makes it difficult to find), but
yeah, it's fair, I'll do it.

File: ports/geckolib/string_cache/lib.rs (hunk 13 above)
Comment (emilio) [inline ports/geckolib/string_cache/lib.rs]: That's true, I removed it. I'll also fill follow-ups to remove the FFI
functions for string comparison and a few other things.

File: ports/geckolib/binding_tools/regen_atoms.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+def gnu_symbolify(source, ident):
+    return "_ZN" + str(len(source.CLASS)) + source.CLASS + str(len(ident)) + ident + "E"
+
+
+def msvc64_symbolify(source, ident):
+    return "?" + ident + "@" + source.CLASS + "@@2PEAV" + source.TYPE + "@@EA"
+
+
+def msvc32_symbolify(source, ident):
+    return "?" + ident + "@" + source.CLASS + "@@2PAV" + source.TYPE + "@@A"
```
Comment (emilio) [inline ports/geckolib/binding_tools/regen_atoms.py]: cc @upsuper, hopefully I got the msvc manglings fine.

**Comment by emilio** [comment]
Left just the atom patches, will submit the content patches after this lands.

@bors-servo: r=bholley

**Comment by emilio** [comment]
@bors-servo: retry

 * #9087 again

File: ports/geckolib/binding_tools/regen_atoms.py (hunk 16 above)
Comment (upsuper) [inline ports/geckolib/binding_tools/regen_atoms.py]: TBH, I have no idea about msvc manglings. Their mangling rule looks complicated, so I disassemble the object file and copy the pattern to here directly.

If the type is just pointers, I guess reusing this is probably fine, but I'm unconfident about that anyway.

**Comment by upsuper** [comment]
Looks like it works fine with MSVC.
